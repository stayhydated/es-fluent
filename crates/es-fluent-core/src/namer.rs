use derive_more::{Debug, Deref, Display};
use heck::ToSnakeCase as _;
use quote::format_ident;

#[derive(Clone, Debug, Deref, Display, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct FluentKey(pub String);

impl FluentKey {
    pub const DELIMITER: &str = "-";
    pub fn new(ftl_name: &syn::Ident, sub_name: &str) -> Self {
        let normalized_name = ftl_name.to_string().to_snake_case();
        if sub_name.is_empty() {
            FluentKey(normalized_name)
        } else {
            FluentKey(format!(
                "{}{}{}",
                normalized_name,
                Self::DELIMITER,
                sub_name
            ))
        }
    }
}

impl quote::ToTokens for FluentKey {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let key_string = &self.0;
        tokens.extend(quote::quote! { #key_string });
    }
}

#[derive(Clone, Debug, Deref, Display, Eq, Hash, PartialEq)]
pub struct FluentDoc(pub String);

impl From<&FluentKey> for FluentDoc {
    fn from(ftl_key: &FluentKey) -> Self {
        FluentDoc(format!("Key = `{}`", *ftl_key))
    }
}

impl quote::ToTokens for FluentDoc {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let doc_string = &self.0;
        tokens.extend(quote::quote! { #doc_string });
    }
}

pub struct UnnamedItem(usize);

impl std::fmt::Display for UnnamedItem {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}{}", Self::UNNAMED_PREFIX, self.0)
    }
}

impl UnnamedItem {
    const UNNAMED_PREFIX: &str = "f";

    pub fn to_ident(&self) -> syn::Ident {
        format_ident!("{}", self.to_string())
    }
}

impl From<usize> for UnnamedItem {
    fn from(index: usize) -> Self {
        Self(index)
    }
}
