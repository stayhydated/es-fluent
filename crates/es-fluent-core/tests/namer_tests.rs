use es_fluent_core::namer::FluentKey;

fn ident(name: &str) -> syn::Ident {
    syn::parse_str::<syn::Ident>(name).expect("valid ident")
}

#[test]
fn fluent_key_formats_snapshot() {
    let cases: Vec<FluentKey> = vec![
        FluentKey::new(&ident("MyStruct"), ""),
        FluentKey::new(&ident("MyStruct"), "Field"),
        FluentKey::new(&ident("HTTPServer"), ""),
        FluentKey::new(&ident("MyEnum"), "Variant"),
        FluentKey::new(&ident("already_snake"), "field_name"),
        FluentKey::new(&ident("X"), ""),
    ];

    insta::assert_ron_snapshot!("fluent_key_formats_snapshot", &cases);
}
