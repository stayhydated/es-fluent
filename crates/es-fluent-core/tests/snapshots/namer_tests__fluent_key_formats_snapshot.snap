---
source: crates/es-fluent-core/tests/namer_tests.rs
expression: "&cases"
---
[
  FluentKey("my_struct"),
  FluentKey("my_struct-Field"),
  FluentKey("http_server"),
  FluentKey("my_enum-Variant"),
  FluentKey("already_snake-field_name"),
  FluentKey("x"),
]
